components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1260,9 +1260,58 @@ where
             SignificantMsg::RaftlogFetched { context, res } => {
                 self.on_raft_log_fetched(context, res);
             }
+            SignificantMsg::EnterForceLeaderState => self.on_enter_force_leader(),
+            SignificantMsg::ExitForceLeaderState => self.on_exit_force_leader(),
         }
     }
 
+    fn on_enter_force_leader(&mut self) {
+        info!(
+            "enter force leader state";
+            "region_id" => self.fsm.region_id(),
+            "peer_id" => self.fsm.peer_id(),
+        );
+        self.fsm.peer.force_leader = true;
+
+        if !self.fsm.peer.is_leader() {
+            // become candidate first to increase term
+            self.fsm.peer.raft_group.raft.become_candidate();
+            // trigger candidate twice to increase term by 2, to avoid there is
+            // a existing leader by accident.
+            self.fsm.peer.raft_group.raft.become_candidate();
+            self.fsm.peer.raft_group.raft.become_leader();
+        }
+        assert!(self.fsm.peer.is_leader());
+        self.fsm.peer.raft_group.raft.set_check_quorum(false);
+
+        // forward commit index
+        self.fsm.peer.raft_group.raft.raft_log.committed =
+            self.fsm.peer.raft_group.raft.raft_log.persisted;
+
+        self.fsm.has_ready = true;
+    }
+
+    fn on_exit_force_leader(&mut self) {
```

Comment:
How is PD detect when to exit the force leader state? if PD relies on the region heatbeat it may possible that the force leader append, commit and apply these `confchange` logs, and send the region heatbeat while other peers may not even have received these `confchange` logs.

Reply:
PD dispatches the recovery plan, and once TiKV finished applying the conf-changes, TiKV will let these regions exit force leader state.

Reply:
What I mean is when the leader applies these `confchange` logs and exit force leader state, other peers may not even have received these `confchange` logs hence still can't elect a new leader.

Reply:
As the conf-change log is applied on the force leader, so after exit force leader state, this peer can elect as a new leader successfully.

Reply:
Oh, I got it